supervision/detection/utils.py

```diff
@@ -1039,3 +1039,75 @@ def cross_product(anchors: np.ndarray, vector: Vector) -> np.ndarray:
     )
     vector_start = np.array([vector.start.x, vector.start.y])
     return np.cross(vector_at_zero, anchors - vector_start)
+
+
+def get_box_intersection(
+    xyxy_1: np.ndarray, xyxy_2: np.ndarray
+) -> Optional[np.ndarray]:
+    overlap_xmin = max(xyxy_1[0], xyxy_2[0])
+    overlap_ymin = max(xyxy_1[1], xyxy_2[1])
+    overlap_xmax = min(xyxy_1[2], xyxy_2[2])
+    overlap_ymax = min(xyxy_1[3], xyxy_2[3])
+
+    if overlap_xmin < overlap_xmax and overlap_ymin < overlap_ymax:
+        return np.array([overlap_xmin, overlap_ymin, overlap_xmax, overlap_ymax])
+    else:
+        return None
+
+
+def get_unit_vector(xy_1: np.ndarray, xy_2: np.ndarray) -> np.ndarray:
+    direction = xy_2 - xy_1
+    magnitude = np.linalg.norm(direction)
+    return direction / magnitude if magnitude > 0 else np.zeros(2)
+
+
+def spread_out_boxes(
+    xyxy: np.ndarray,
+    max_iterations: int = 100,
+    force_multiplier: float = 0.03,
+) -> np.ndarray:
+    """
+    Spread out boxes that overlap with each other.
+
+    Args:
+        xyxy: Numpy array of shape (N, 4) where N is the number of boxes.
+        max_iterations: Maximum number of iterations to run the algorithm for.
+        force_multiplier: Multiplier to scale the force vectors by. Similar to
+            learning rate in gradient descent.
+    """
+    if len(xyxy) == 0:
+        return xyxy
+
+    xyxy_padded = pad_boxes(xyxy, px=1)
+    for _ in range(max_iterations):
+        # NxN
+        iou = box_iou_batch(xyxy_padded, xyxy_padded)
+        np.fill_diagonal(iou, 0)
+        if np.all(iou == 0):
+            break
+
+        overlap_mask = iou > 0
+
+        # Nx2
+        centers = (xyxy_padded[:, :2] + xyxy_padded[:, 2:]) / 2
+
+        # NxNx2
+        delta_centers = centers[:, np.newaxis, :] - centers[np.newaxis, :, :]
+        delta_centers *= overlap_mask[:, :, np.newaxis]
+
+        # Nx2
+        force_vectors = np.sum(delta_centers, axis=1)
```

Comment:
I was playing with some visualization ideas and noticed some strange behavior in "smart mode." Small changes in box position can lead to large changes in label position. As a result, the overall effect can feel choppy.

I'm afraid math breaks here. :/ 

The movement vector should be proportional to the IoU - the more the boxes overlap, the more we need to pull them apart. Instead, the force with which we push them apart is proportional to the distance between their centers. If the centers are further apart, the boxes are pushed apart more strongly.